.github/workflows/ci_linux_arm64_clang.yml

```diff
@@ -0,0 +1,63 @@
+# Copyright 2024 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+name: CI - Linux arm64 clang
+
+on:
+  schedule:
+    # Weekday mornings at 09:15 UTC = 01:15 PST (UTC - 8).
+    - cron: "15 9 * * 1-5"
+  workflow_dispatch:
+
+concurrency:
+  # A PR number if a pull request and otherwise the commit hash. This cancels
+  # queued and in-progress runs for the same PR (presubmit) or commit
+  # (postsubmit). The workflow name is prepended to avoid conflicts between
+  # different workflows.
+  group: ${{ github.workflow }}-${{ github.event.number || github.sha }}
+  cancel-in-progress: true
+
+jobs:
+  setup:
+    uses: ./.github/workflows/setup.yml
+
+  # TODO: Switch to GitHub-hosted arm64 runners when available?
+  # TODO: Switch runs-on labels to use different (non-GCP) self-hosted runners?
+  linux_arm64_clang:
+    needs: setup
+    if: contains(fromJson(needs.setup.outputs.enabled-jobs), 'build_test_all_arm64')
+    runs-on:
+      - self-hosted # must come first
+      - runner-group=${{ needs.setup.outputs.runner-group }}
+      - environment=${{ needs.setup.outputs.runner-env }}
+      - arm64
+      - os-family=Linux
+    env:
+      BUILD_DIR: build-arm64
+    steps:
+      - name: "Checking out repository"
+        uses: actions/checkout@v4.1.7
+        with:
+          submodules: true
+      - name: "Building IREE"
+        env:
+          IREE_WRITE_REMOTE_CCACHE: ${{ needs.setup.outputs.write-caches }}
+        run: |
+          ./build_tools/github_actions/docker_run.sh \
+            --env "IREE_CCACHE_GCP_TOKEN=$(gcloud auth application-default print-access-token)" \
```

Comment:
I am guessing that this is setting up some CCACHE magic specific to GCP? Something that wouldn't work on self-hosted runners outside of GCP?

Reply:
Yep. I just copied the previous job to this new file with only minor edits. The other jobs aren't using any caching at the moment, so they are pretty slow. If you switch this to your own runner, you might want to remove/adjust the Docker code and use a different (or no) cache mechanism.

We've experimented with different caching options:
* Self-hosted persistent runners with local ccache
* Self-hosted ephemeral runners with cloud ccache (e.g. GCP/GCS)
* GitHub-hosted runners with actions caches: https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows#usage-limits-and-eviction-policy

Local caches on persistent runners are simplest :)

Cloud caches near the build machines (e.g. GCP using GCS) should have good network latency and bandwidth, as well as growable sizes and configurable TTLs / eviction policies.

GitHub's caches are pretty limited - 10GB across all caches in the repository before old entries start getting evicted. I remember 2-3GB caches being slow to upload/download, and the compiler typically needs ~3GB to get the 95%+ cache hit rates that speed up the build enough. In my experiments, if the cache hit rate got to ~80% then the build time would jump up substantially (it's been a while, but IIRC it was like 15 minutes --> 40 minutes).

Reply:
> 95%+ cache hit rates 

Presumably for a clean build? I've never hit such numbers for incremental builds. This is beside the point, I'm just curious :)

Reply:
Hmmm... Not sure for arm64 with clang. Many of my experiments were for Windows with MSVC.

arm64 from last night:
* 1h30m first build (cold cache): https://github.com/iree-org/iree/actions/runs/10049822259/job/27776581816#step:4:10448
    ```
    + ccache --show-stats
    Cacheable calls:   5709 / 5729 (99.65%)
      Hits:             660 / 5709 (11.56%)
        Direct:         144 /  660 (21.82%)
        Preprocessed:   516 /  660 (78.18%)
      Misses:          5049 / 5709 (88.44%)
    Uncacheable calls:   20 / 5729 ( 0.35%)
    Local storage:
      Cache size (GB): 0.00 / 5.00 ( 0.00%)
      Hits:               0
      Misses:             0
    Remote storage:
      Hits:             660 / 5716 (11.55%)
      Misses:          5056 / 5716 (88.45%)
      Errors:          5233
    ```
* 50m second build (warm cache): https://github.com/iree-org/iree/actions/runs/10056339628/job/27794936386#step:4:10453
    ```
    + ccache --show-stats
    Cacheable calls:   5709 / 5729 (99.65%)
      Hits:            3428 / 5709 (60.05%)
        Direct:         931 / 3428 (27.16%)
        Preprocessed:  2497 / 3428 (72.84%)
      Misses:          2281 / 5709 (39.95%)
    Uncacheable calls:   20 / 5729 ( 0.35%)
    Local storage:
      Cache size (GB): 0.00 / 5.00 ( 0.00%)
      Hits:               0
      Misses:             0
    Remote storage:
      Hits:            3428 / 5715 (59.98%)
      Misses:          2287 / 5715 (40.02%)
      Errors:          4379
    ```


Compare that with linux x86 clang (on a VM with more cores):
* 13m build after an LLVM integrate (so cold cache): https://github.com/iree-org/iree/actions/runs/10012178032/job/27677127470#step:4:10933
    ```
    + ccache --show-stats
    Cacheable calls:   5802 / 5822 (99.66%)
      Hits:            2335 / 5802 (40.24%)
        Direct:         824 / 2335 (35.29%)
        Preprocessed:  1511 / 2335 (64.71%)
      Misses:          3467 / 5802 (59.76%)
    Uncacheable calls:   20 / 5822 ( 0.34%)
    Local storage:
      Cache size (GB): 0.00 / 5.00 ( 0.00%)
      Hits:               0
      Misses:             0
    Remote storage:
      Hits:            2335 / 5810 (40.19%)
      Misses:          3475 / 5810 (59.81%)
      Errors:          4497
    ```
* 5m build after a no-op (so warm cache): https://github.com/iree-org/iree/actions/runs/10012585025/job/27678397517#step:4:10933
    ```
    + ccache --show-stats
    Cacheable calls:   5802 / 5822 (99.66%)
      Hits:            5801 / 5802 (99.98%)
        Direct:        1006 / 5801 (17.34%)
        Preprocessed:  4795 / 5801 (82.66%)
      Misses:             1 / 5802 ( 0.02%)
    Uncacheable calls:   20 / 5822 ( 0.34%)
    Local storage:
      Cache size (GB): 0.00 / 5.00 ( 0.00%)
      Hits:               0
      Misses:             0
    Remote storage:
      Hits:            5801 / 5810 (99.85%)
      Misses:             9 / 5810 ( 0.15%)
      Errors:          4439
    ```